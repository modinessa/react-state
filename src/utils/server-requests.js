//const url = 'http://localhost:3000';
const url = '';

export function getUsers() {
  return fetch(`${url}/community`, {
    method: 'GET',
  });
}

export function getUser(id) {
	return fetch(`/community/${id}`);
}

export function subscribe(email) {
  return fetch(`${url}/subscribe`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      email: `${email}`,
    }),
  });
}

export function unsubscribe() {
  return fetch(`${url}/unsubscribe`, { method: 'POST' });
}
