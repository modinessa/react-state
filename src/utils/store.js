import { configureStore } from "@reduxjs/toolkit";
import usersReducer from "../components/reducer";

export const store = configureStore({
	reducer: {
		users: usersReducer,
	}
});