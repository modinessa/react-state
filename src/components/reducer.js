import { createAsyncThunk, createSlice} from "@reduxjs/toolkit";
import { getUser, getUsers } from "../utils/server-requests";

export const fetchUsers = createAsyncThunk(
  'users/fetchUsers',
  async () => {
		return getUsers()
			.then((response) => response.json())
 }
);

export const fetchUser = createAsyncThunk(
	'users/fetchUser',
	async (id) => {
		return getUser(id)
			.then((response) => response.json());
	}
)

//const isSubscribed = localStorage.getItem('isSubscribed') === 'true';

const isHide = localStorage.getItem('isHide') === 'true';

export const usersSlice = createSlice({
	name: "users",
	initialState: {
			isHide,
			users: [],
			isSubscribed: false,
			isSubmitting: false,
			user: null,
			userNotFound: false,
	},
	extraReducers: {
      [fetchUsers.fulfilled]: (state, {payload}) => {
        state.users = payload;
      },
			[fetchUser.fulfilled]: (state, {payload}) => {
				state.user = payload;
				state.userNotFound = false;
			},
			[fetchUser.rejected]: (state) => {
				state.user = null;
				state.userNotFound = true;
			},
  },
	reducers: {
		setIsHide: (state, {payload}) => {
			state.isHide = payload;
			localStorage.setItem('isHide', payload);
		},
		setIsSubscribed: (state, {payload}) => {
			state.isSubscribed = payload;
			localStorage.setItem('isSubscribed', payload);
		},
		setIsSubmitting: (state, {payload}) => {
			state.isSubmitting = payload;
		},
	}
});

export const {
	setIsHide,
	setIsSubscribed, 
	setIsSubmitting,
} = usersSlice.actions;
export default usersSlice.reducer;
