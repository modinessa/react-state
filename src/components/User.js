import {useParams, useNavigate} from "react-router-dom";
import {useSelector, useDispatch} from "react-redux";
import * as constants from "../constants/constants";
import {fetchUser} from "./reducer";
import { useEffect } from "react";

export function User() {
	const {id} = useParams();
	const dispatch = useDispatch();
	const navigate = useNavigate();

	console.log(id);

	useEffect(() => {
		dispatch(fetchUser(id));
	});

	const {user, userNotFound} = useSelector((state) => state.users);
	if (userNotFound) {
		navigate('/not-found');
	}

	return user ? (
		<div className="app-section">
			<div key={user.id} className={constants.USER_CARD_CLASS}>
				<img src={user.avatar} className="app-section__member-img"
						alt="User"/>
				<p className="app-section__member-review">
					{constants.USER_REVIEW}
				</p>
				<div className="app-section__member-name">
					{user.firstName} {user.lastName}
				</div>
				<div className="app-section__member-position">
					{user.position}
				</div>
			</div>
		</div>
	) : <></>;
}