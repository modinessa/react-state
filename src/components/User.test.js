import renderer from 'react-test-renderer';
import {Provider} from "react-redux";
import {rest} from 'msw';
import {setupServer} from 'msw/node';
import {BrowserRouter, Route, Routes} from 'react-router-dom'; 

import {store} from "../utils/store";
import {User} from './User';
import {fetchUser} from './reducer';

const TEST_USER_ID = 'fake_id';
const TEST_USER = {
	id: TEST_USER_ID,
	firstName: 'first',
	lastName: 'last',
	position: 'test',
	avatar: 'fake_url',
};

const mockServer = setupServer(
	rest.get(`/community/${TEST_USER_ID}`,
	  (req, res, ctx) => res(ctx.json(TEST_USER))),
)

describe('User', () => {
	beforeAll(() => {
		mockServer.listen();
		window.history.pushState({}, '', `/community/${TEST_USER_ID}`);
	});

	afterAll(() => {
		mockServer.close();
	});

	it('renders component with user information', async () => {
		await store.dispatch(fetchUser(TEST_USER_ID));

		const user = renderer.create(
			<Provider store={store}>
				<BrowserRouter>
					<Routes>
						<Route path="/community/:id" element={<User/>}/>
					</Routes>
				</BrowserRouter>
			</Provider>);

			expect(user).toMatchSnapshot();
	});
});
