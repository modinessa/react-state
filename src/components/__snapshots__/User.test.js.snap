// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User renders component with user information 1`] = `
<div
  className="app-section"
>
  <div
    className="app-section__member-card"
  >
    <img
      alt="User"
      className="app-section__member-img"
      src="fake_url"
    />
    <p
      className="app-section__member-review"
    >
      PLACEHOLDER: Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do incididunt ut labore et dolor.
    </p>
    <div
      className="app-section__member-name"
    >
      first
       
      last
    </div>
    <div
      className="app-section__member-position"
    >
      test
    </div>
  </div>
</div>
`;
